### PR TITLE
feat: add protectConnectionStrings option COMPASS-6066

### PR DESCRIPTION
**Pull request by addaleax**
### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-maybe-protect-connection-string/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import preferences from 'compass-preferences-model';
+import { redactConnectionString } from 'mongodb-connection-string-url';
+
+export function maybeProtectConnectionString(connectionString: string): string {
+  if (preferences.getPreferences().protectConnectionStrings) {
+    return redactConnectionString(connectionString);
+  }
+  return connectionString;
+}
```
Comment (addaleax) [inline packages/compass-maybe-protect-connection-string/src/index.ts:9]: Obviously this is a bunch of overhead for a single function, but at the same time: :woman_shrugging: ?

This might have fit into compass-utils but that would have resulted in a circular dependency, and it’s not like it really hurts to have another package with our current monorepo setup.

File: packages/compass-e2e-tests/helpers/commands/set-feature.ts [hunk 2]
```diff
@@ -8,9 +8,11 @@ export async function setFeature(
...
     async (_name, _value) => {
       // eslint-disable-next-line @typescript-eslint/no-var-requires
-      await require('electron').ipcRenderer.invoke('compass:save-preferences', {
-        [_name]: _value,
-      });
+      await require('compass-preferences-model').preferencesAccess.savePreferences(
+        {
+          [_name]: _value,
+        }
+      );
```
Comment (addaleax) [inline packages/compass-e2e-tests/helpers/commands/set-feature.ts:15]: Small drive-by fix to decouple this from the actual implementation of `savePreferences`

File: packages/compass-e2e-tests/tests/import-export-connections.test.ts [hunk 3]
```diff
@@ -119,6 +127,7 @@ describe('Connection Import / Export', function () {
         await runCompassOnce([
           `--import-connections=${file}`,
           ...passphraseArgs,
+          '--trackUsageStatistics',
```
Comment (addaleax) [inline packages/compass-e2e-tests/tests/import-export-connections.test.ts:130]: The logging test sets `trackUsageStatistics: true`, which I think is what made this pass before, but makes it harder to work on these tests in isolation. Just passing the flag on the command line ensures that telemetry is always enabled.

**Review by addaleax — COMMENTED** [review]
(Split out adding this to the settings UI into COMPASS-6262 since there might be conflicts with the `readOnly` PR otherwise)

File: packages/compass-e2e-tests/helpers/commands/set-feature.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-e2e-tests/helpers/commands/set-feature.ts:15]: Doesn't this rely on the fact that we are running the application inside the monorepo currently so the module can be resolved outside the bundled application code? In the real app this will fail with module not found, and we have plans to move the app outside the monorepo specifically so we can test that code is bundled with all required dependencies. 

If we don't want to rely on implementation details, I'd suggest exposing preferences globally in the renderer code itself (maybe only in test / dev mode), which is also handy when developing

File: packages/connection-form/src/components/connect-form.tsx [hunk 4]
```diff
@@ -27,6 +27,7 @@ import ConnectFormActions from './connect-form-actions';
 import { useConnectForm } from '../hooks/use-connect-form';
 import { validateConnectionOptionsErrors } from '../utils/validation';
 import SaveConnectionModal from './save-connection-modal';
+import { usePreference } from 'compass-preferences-model';
```
Comment (Anemy) [inline packages/connection-form/src/components/connect-form.tsx:30]: Does this prohibit this package from being used in a non-Compass setting? Thinking about if we want to use it in VSCode or other web apps down the line.

**Review by gribnoysup — APPROVED** [review]
Looks good to me!

File: packages/compass-connections/package.json [hunk 5]
```diff
@@ -66,6 +66,7 @@
   },
   "devDependencies": {
     "@mongodb-js/compass-components": "^1.3.0",
+    "@mongodb-js/compass-maybe-protect-connection-string": "^0.1.0",
```
Comment (alenakhineika) [inline packages/compass-connections/package.json:69]: Maybe 😄👏
